jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -954,6 +954,269 @@ ecma_promise_then (ecma_value_t promise, /**< the promise which call 'then' */
+ *         ES2020 25.6.5.3.2 steps 2-7
+ *
+ * @return ecma value
+ */
+static ecma_value_t
+ecma_promise_than_catch_finally_helper (ecma_value_t function_obj) /**< the function itself */
+{
+  /* 2. */
+  ecma_object_t *func_obj_p = ecma_get_object_from_value (function_obj);
+  ecma_promise_finally_function_t *finally_func_obj = (ecma_promise_finally_function_t *) func_obj_p;
+
+  /* 3. */
+  JERRY_ASSERT (ecma_op_is_callable (finally_func_obj->on_finally));
+
+  /* 4. */
+  ecma_value_t result = ecma_op_function_call (ecma_get_object_from_value (finally_func_obj->on_finally),
+                                               ECMA_VALUE_UNDEFINED,
+                                               NULL,
+                                               0);
+
+  if (ECMA_IS_VALUE_ERROR (result))
+  {
+    return result;
+  }
+
+  /* 6. */
+  JERRY_ASSERT (ecma_is_constructor (finally_func_obj->constructor));
+
+  /* 7. */
+  ecma_value_t promise = ecma_promise_reject_or_resolve (finally_func_obj->constructor, result, true);
+
+  ecma_free_value (result);
+
+  return promise;
+} /* ecma_promise_than_catch_finally_helper */
+
+/**
+ * Definition of Then Finally Function
+ *
+ * See also:
+ *         ES2020 25.6.5.3.1
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_promise_then_finally_cb (const ecma_value_t function_obj, /**< the function itself */
+                              const ecma_value_t this_val, /**< this_arg of the function */
+                              const ecma_value_t args_p[], /**< argument list */
+                              const ecma_length_t args_count) /**< argument number */
+{
+  JERRY_UNUSED_3 (this_val, args_p, args_count);
+
+  /* 7. */
+  ecma_value_t promise = ecma_promise_than_catch_finally_helper (function_obj);
+
+  if (ECMA_IS_VALUE_ERROR (promise))
+  {
+    return promise;
+  }
+
+  /* 8. */
+  ecma_object_t *thunk_obj_p;
+  thunk_obj_p = ecma_create_object (ecma_builtin_get (ECMA_BUILTIN_ID_FUNCTION_PROTOTYPE),
+                                    sizeof (ecma_promise_value_thunk_t),
+                                    ECMA_OBJECT_TYPE_EXTERNAL_FUNCTION);
+
+  ecma_promise_value_thunk_t *thunk_func_obj = (ecma_promise_value_thunk_t *) thunk_obj_p;
+  thunk_func_obj->header.u.external_handler_cb = ecma_value_thunk_helper_cb;
+
+  JERRY_ASSERT (args_count > 0);
+  thunk_func_obj->value = ecma_copy_value_if_not_object (args_p[0]);
+
+  /* 9. */
+  ecma_value_t value_thunk = ecma_make_object_value (thunk_obj_p);
+  ecma_value_t ret_value = ecma_op_invoke_by_magic_id (promise, LIT_MAGIC_STRING_THEN, &value_thunk, 1);
+
+  ecma_free_value (promise);
+  ecma_deref_object (thunk_obj_p);
+
+  return ret_value;
+} /* ecma_promise_then_finally_cb */
+
+/**
+ * Definition of Catch Finally Function
+ *
+ * See also:
+ *         ES2020 25.6.5.3.2
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_promise_catch_finally_cb (const ecma_value_t function_obj, /**< the function itself */
```

Comment:
Would it make sense to extract the common parts of this and the previous method into a handler method? If I read it correctly the only difference is the `external_handler_cb`.